logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-            }
-            if (ResolvedType == null) {
-                subVar.removeResolvedType(label);
-                if (subVar.resolvedTypes().isEmpty()) subVar.setSatisfiable(false);
-            }
-        }
-    }
+        Map<Reference, Set<Label>> resolvedLabels = runTraversalEngine(traversalConstructor);
+        if (resolvedLabels.isEmpty()) throw GraknException.of(SCHEMATICALLY_UNSATISFIABLE_CONJUNCTION, conjunction);
+        resolvedLabels.forEach((ref, labels) -> traversalConstructor.getVariable(ref).addResolvedTypes(labels));
+        findVariablesWithNoHints(conjunction);
 
-    private void addInferredSubLabels(TypeVariable variable, Set<Label> resolvedLabels) {
-        variable.addResolvedTypes(resolvedLabels);
+        return conjunction;
     }
 
-    private void addInferredIsaLabels(ThingVariable variable, Set<Label> resolveLabels) {
-        //TODO: use .getType(label) once ConceptManager can handle labels
-        resolveLabels.removeIf(label -> traversalEng.graph().schema().getType(label).isAbstract());
-        variable.addResolvedTypes(resolveLabels);
+    private void findVariablesWithNoHints(Conjunction conjunction) {
+        long numOfTypes = traversalEng.graph().schema().stats().thingTypeCount();
+        long numOfConcreteTypes = traversalEng.graph().schema().stats().concreteThingTypeCount();
+        iterate(conjunction.variables()).filter(variable -> (
+                variable.isType() && variable.resolvedTypes().size() == numOfTypes ||
+                        variable.isThing() && variable.resolvedTypes().size() == numOfConcreteTypes
+        )).forEachRemaining(Variable::clearResolvedTypes);
     }
 
-    private Map<Reference, Set<Label>> retrieveResolveTypes(Set<Variable> resolveVars) {
-        Conjunction resolveVariableConjunction = new Conjunction(resolveVars, Collections.emptySet());
-        resolveVariableConjunction = resolveLabels(resolveVariableConjunction);
-        return logicCache.resolver().get(resolveVariableConjunction, conjunction -> {
+    private Map<Reference, Set<Label>> runTraversalEngine(TraversalConstructor traversalConstructor) {
+        return logicCache.resolver().get(traversalConstructor.resolverTraversal(), traversal -> {
             Map<Reference, Set<Label>> mapping = new HashMap<>();
-            traversalEng.iterator(conjunction.traversal()).forEachRemaining(
+            traversalEng.iterator(traversal).forEachRemaining(
                     result -> result.forEach((ref, vertex) -> {
                         mapping.putIfAbsent(ref, new HashSet<>());
-                        mapping.get(ref).add(Label.of(vertex.asType().label(), vertex.asType().scope()));
+                        assert vertex.isType();
+                        if (!(vertex.asType().isAbstract() && traversalConstructor.getVariable(ref).isThing()))
+                            mapping.get(ref).add(Label.of(vertex.asType().label(), vertex.asType().scope()));
```

Comment:
We could have put the results directly into the `conjunction`. But we put it into a Map and then place the labels from the Map to the `conjunction` so we can cache the Map for later use.